File: 02-java-spring-boot/src/main/java/com/dprev/checkout/RunnerEstudos.java
package com.dprev.checkout;

import com.dprev.checkout.domain.cascade.CartaoCorporativoPremium;
import com.dprev.checkout.domain.strategy.CartaoCreditoPagamento;
import com.dprev.checkout.domain.strategy.PixPagamento;
import com.dprev.checkout.domain.strategy.ValeRefeicaoPagamento;
import com.dprev.checkout.service.CheckoutService;
import org.springframework.boot.CommandLineRunner;
import org.springframework.stereotype.Component;

@Component
public class RunnerEstudos implements CommandLineRunner {

    private final CheckoutService checkoutService;
    private final PixPagamento pix;
    private final CartaoCreditoPagamento cartaoCredito;
    private final CartaoCorporativoPremium cartaoCorp;
    private final ValeRefeicaoPagamento vr;

    public RunnerEstudos(CheckoutService checkoutService,
                         PixPagamento pix,
                         CartaoCreditoPagamento cartaoCredito,
                         CartaoCorporativoPremium cartaoCorp,
                         ValeRefeicaoPagamento vr) {
        this.checkoutService = checkoutService;
        this.pix = pix;
        this.cartaoCredito = cartaoCredito;
        this.cartaoCorp = cartaoCorp;
        this.vr = vr;
    }

    @Override
    public void run(String... args) throws Exception {
        System.out.println("\n🧪 ========================================================");
        System.out.println("🧪 INICIANDO SUITE DE TESTES AUTOMATIZADOS (SPRING BOOT)");
        System.out.println("🧪 ========================================================\n");

        System.out.println("🔹 [CENÁRIO 1] Processamento Polimórfico Padrão...");
        checkoutService.finalizarPedido(100.0, pix, "fabiano@email.com");
        checkoutService.finalizarPedido(250.50, cartaoCredito, "fabiano@email.com");

        System.out.println("🔹 [CENÁRIO 2] Extensões Complexas e Contratos do Cartão Premium...");
        checkoutService.finalizarPedido(5000.0, cartaoCorp, "diretoria@dprev.com");
        cartaoCorp.generateLogAuditoria(5000.0);
        cartaoCorp.processarTransacaoInternacional("USD");
        cartaoCorp.estornar(5000.0);

        System.out.println("\n🔹 [CENÁRIO 3] Análise de Quebra do Liskov Substitution Principle (LSP)...");
        checkoutService.finalizarPedido(50.0, vr, "fabiano@email.com");

        try {
            checkoutService.finalizarPedido(350.0, vr, "fabiano@email.com");
        } catch (IllegalArgumentException e) {
            System.out.println("[ALERTA ERRO] LSP Quebrado sob análise: " + e.getMessage());
            System.out.println("✅ SUCESSO AUTOMATIZADO: O Spring tratou a exceção do polimorfismo restrito.");
        }

        System.out.println("\n🚀 ========================================================");
        System.out.println("🚀 TESTES CONCLUÍDOS. O SERVIDOR CONTINUA ATIVO NA PORTA 8080.");
        System.out.println("🚀 ========================================================\n");
    }
}

File: 04-kotlin-spring-boot/src/main/kotlin/com/dprev/checkout/RunnerEstudos.kt
package com.dprev.checkout

import com.dprev.checkout.domain.strategy.PixPagamento
import com.dprev.checkout.domain.strategy.CartaoCreditoPagamento
import com.dprev.checkout.domain.strategy.BoletoPagamento
import com.dprev.checkout.domain.strategy.ValeRefeicaoPagamento
import com.dprev.checkout.domain.cascade.CartaoCorporativoPremium
import com.dprev.checkout.service.CheckoutService
import org.springframework.boot.CommandLineRunner
import org.springframework.stereotype.Component

@Component
public class RunnerEstudos (
    private var checkout : CheckoutService,
    private var pix: PixPagamento,
    private var cartaoCredito: CartaoCreditoPagamento,
    private var shadowBoleto: BoletoPagamento,
    private var vr: ValeRefeicaoPagamento,
    private var cartaoCorp: CartaoCorporativoPremium
): CommandLineRunner {

    // Injeção de dependência pelo construtor do Spring Boot
    public fun RunnerEstudos(checkout: CheckoutService , pix: PixPagamento , cartaoCredito: CartaoCreditoPagamento ,
                             shadowBoleto: BoletoPagamento , vr :ValeRefeicaoPagamento , cartaoCorp: CartaoCorporativoPremium ) {
        this.checkout = checkout;
        this.pix = pix;
        this.cartaoCredito = cartaoCredito;
        this.shadowBoleto = shadowBoleto;
        this.vr = vr;
        this.cartaoCorp = cartaoCorp;
    }

    @Override
    public override fun run(vararg args: String) {
        println("\n=== TESTANDO ESTRUTURA ORIGINAL VIA SPRING ===");
        // [🟩O] e [🟨L] - POLIMORFISMO: O checkout consome as estratégias estavelmente
        // via contrato 'MetodoPagamento', sem necessidade de conhecer suas especificidades.
        checkout.finalizarPedido(100.0, pix, "fabiano@email.com");
        checkout.finalizarPedido(250.50, cartaoCredito, "fabiano@email.com");
        checkout.finalizarPedido(50.0, shadowBoleto, "fabiano@email.com");

        println("=== TESTANDO CONCEITOS DE CASCATA (EXTENDS & IMPLEMENTS) ===");
        checkout.finalizarPedido(5000.0, cartaoCorp, "diretoria@dprev.com");

        println("Verificando rastreabilidade do cartão corporativo:");
        cartaoCorp.generateLogAuditoria(5000.0);
        cartaoCorp.processarTransacaoInternacional("USD");
        cartaoCorp.estornar(5000.0);

        System.out.println("\n=== TESTANDO VIOLAÇÃO DO LSP (ESTUDO DE CASO) ===");
        checkout.finalizarPedido(50.0, vr, "fabiano@email.com")

        // [🟨L] - LISKOV SUBSTITUTION PRINCIPLE (VIOLAÇÃO): O código cliente espera comportamento uniforme.
        // Lançar uma exceção de validação oculta em uma subclasse quebra as expectativas do polimorfismo puro.
        try {
            checkout.finalizarPedido(350.0, vr, "fabiano@email.com")
        } catch (e:IllegalArgumentException ) {
            // CORRIGIDO: Sintaxe de interpolação do Kotlin trocada pela concatenação do Java e uso de getMessage()
            println("[ALERTA ERRO] LSP Quebrado sob análise: ${e.message}")
            println("O código cliente não deve ser forçado a prever regras restritivas exclusivas de um subtipo.")
        }
    }
}